### DISCUSSION TITLE
On updating edge attributes

### DISCUSSION THREAD

**Discussion by leonardtschora** [pyg-team/pytorch_geometric#5329] [Q&A] [answered]
Hi everyone,

I am struggling to understand how exactly Edge Feature updating works in the pytorch framework.
More precisely, here is a list of questions that I hope are sufficiently clear for being answered.

1. As I understand, it is possible to update edge attributes using the 'edge_update' function of a Conv layer. Using this function works similarly than the 'forward' function. Can you confirm that, up to this date, only the [GatConv layer](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html?highlight=GatConv#torch_geometric.nn.conv.GATConv) has this function implemented? (on the latest version, 2.1.0).
2. In the [GatConv paper](https://arxiv.org/abs/1710.10903), the edge attribute updating procedure is not put forward. Moreover, it seems to be missing from the [GatConv layer documentation](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html?highlight=GatConv#torch_geometric.nn.conv.GATConv). Are the edge attributes replaced by the aggregation of attention coefficients?
3. Lastly, I feel that the [ResGatedGraphConv layer](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html?highlight=GatConv#torch_geometric.nn.conv.ResGatedGraphConv) should support edge attribute updating. Indeed, authors of the [paper](https://arxiv.org/abs/1711.07553) mention edge gates that seems to update  the edge attributes. This is confirmed in this [article review](https://arxiv.org/abs/2003.00982). Is there any particular reason why this functionality is not available?

I hope I have been clear enough. I thank you by advance for your help, any input would be very appreciated.

**Comment by rusty1s** [comment, marked as answer]
1. Yes, currently only `GATConv` makes use of it. The tests have some additional examples, see [here](https://github.com/pyg-team/pytorch_geometric/blob/master/test/nn/conv/test_message_passing.py#L243).
2. In the current design, the edge features are used to derive at the attention coefficients between source and destination nodes.
3. I am happy to include it there if you want to make a PR :)

Let me know if this resolves your concerns.

**Reply by leonardtschora** [reply]
Thanks for the answer! Just to be sure, the `GATConv` layer does NOT updates edges embeddings, it only computes them, right?

**Reply by rusty1s** [reply]
That is correct, if you want to update them you would have to return them as part of `forward` and just them in later layers.

**Reply by leonardtschora** [reply]
Alright, thank you very much!
